src/sorting_network.cpp

```diff
@@ -0,0 +1,199 @@
+/*********************                                                        */
+/*! \file sorting_network.cpp
+** \verbatim
+** Top contributors (to current version):
+**   Makai Mann
+** This file is part of the smt-switch project.
+** Copyright (c) 2020 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief Implements a symbolic sorting network for boolean-sorted variables.
+**
+**/
+
+#include "sorting_network.h"
+
+#include <assert.h>
+
+#include "exceptions.h"
+
+namespace smt {
+
+// implementation based on SortingNetwork in the predecessor, CoSA:
+// https://github.com/cristian-mattarei/CoSA/blob/141be4b23f4012c6ad5676907d7c211ae2b6614a/cosa/utils/formula_mngm.py#L198
+
+TermVec SortingNetwork::sorting_network(const TermVec & unsorted) const
+{
+  if (unsorted.empty())
+  {
+    return {};
+  }
+
+  // check that all the terms in unsorted are boolean sorted
+  // for sort aliasing solvers, best to compare to the object
+  // rather than rely on the SortKind
+  Sort boolsort = solver_->make_sort(BOOL);
+  Sort sort;
+  for (const auto & tt : unsorted)
+  {
+    sort = tt->get_sort();
+    if (tt->get_sort() != boolsort)
+    {
+      throw IncorrectUsageException("Expected all boolean sorts but got "
+                                    + tt->to_string() + ":"
+                                    + sort->to_string());
+    }
+  }
+
+  return sorting_network_rec(unsorted);
+}
+
+TermVec SortingNetwork::sorting_network_rec(const TermVec & unsorted) const
+{
+  size_t len = unsorted.size();
+  if (len == 1)
```

Comment:
```suggestion
  if (len == 0 || len == 1)
```
Because I fear we might get an infinite loop if `len=0`? You check this in `sorting_network` but what if you keep dividing by 2 and then reach `0`?